UM/Math/Matrix.py

```diff
@@ -140,6 +168,177 @@ def setByRotationAxis(self, angle, direction, point = None):
+                ax = math.atan2( M[i, j],  M[i, k])
+                ay = math.atan2( sy,       M[i, i])
+                az = math.atan2( M[j, i], -M[k, i])
+            else:
+                ax = math.atan2(-M[j, k],  M[j, j])
+                ay = math.atan2( sy,       M[i, i])
+                az = 0.0
+        else:
+            cy = math.sqrt(M[i, i] * M[i, i] + M[j, i] * M[j, i])
+            if cy > self._EPS:
+                ax = math.atan2( M[k, j],  M[k, k])
+                ay = math.atan2(-M[k, i],  cy)
+                az = math.atan2( M[j, i],  M[i, i])
+            else:
+                ax = math.atan2(-M[j, k],  M[j, j])
+                ay = math.atan2(-M[k, i],  cy)
+                az = 0.0
+
+        if parity:
+            ax, ay, az = -ax, -ay, -az
+        if frame:
+            ax, az = az, ax
+        return Vector(ax, ay, az)
+
+
+    def setByEuler(self, ai, aj, ak, axes = "sxyz"):
+        """Return homogeneous rotation matrix from Euler angles and axis sequence.
+
+        ai, aj, ak : Euler's roll, pitch and yaw angles
+        axes : One of 24 axis sequences as string or encoded tuple
+
+        >>> R = euler_matrix(1, 2, 3, 'syxz')
+        >>> numpy.allclose(numpy.sum(R[0]), -1.34786452)
+        True
+        >>> R = euler_matrix(1, 2, 3, (0, 1, 0, 1))
+        >>> numpy.allclose(numpy.sum(R[0]), -0.383436184)
+        True
+        >>> ai, aj, ak = (4*math.pi) * (numpy.random.random(3) - 0.5)
+        >>> for axes in _AXES2TUPLE.keys():
+        ...    R = euler_matrix(ai, aj, ak, axes)
+        >>> for axes in _TUPLE2AXES.keys():
+        ...    R = euler_matrix(ai, aj, ak, axes)
+
+        """
+        try:
+            firstaxis, parity, repetition, frame = self._AXES2TUPLE[axes]
+        except (AttributeError, KeyError):
+            self._TUPLE2AXES[axes]  # validation
+            firstaxis, parity, repetition, frame = axes
+        i = firstaxis
+        j = self._NEXT_AXIS[i + parity]
+        k = self._NEXT_AXIS[i - parity + 1]
+
+        if frame:
+            ai, ak = ak, ai
+        if parity:
+            ai, aj, ak = -ai, -aj, -ak
+
+        si, sj, sk = math.sin(ai), math.sin(aj), math.sin(ak)
+        ci, cj, ck = math.cos(ai), math.cos(aj), math.cos(ak)
+        cc, cs = ci * ck, ci * sk
+        sc, ss = si * ck, si * sk
+
+        M = numpy.identity(4)
+        if repetition:
+            M[i, i] = cj
+            M[i, j] = sj * si
+            M[i, k] = sj * ci
+            M[j, i] = sj * sk
+            M[j, j] = -cj * ss + cc
+            M[j, k] = -cj * cs - sc
+            M[k, i] = -sj * ck
+            M[k, j] = cj * sc + cs
+            M[k, k] = cj * cc - ss
+        else:
+            M[i, i] = cj * ck
+            M[i, j] = sj * sc - cs
+            M[i, k] = sj * cc + ss
+            M[j, i] = cj * sk
+            M[j, j] = sj * ss + cc
+            M[j, k] = sj * cs - sc
+            M[k, i] = -sj
+            M[k, j] = cj * si
+            M[k, k] = cj * ci
```

Comment:
I trust the validity of the unit test they have provided (and the fact it was used in peer-reviewed academic research). I've also never looked in the euler lib code and I trust those results ;)